Content.Server/GameObjects/Components/Kitchen/ReagentGrinderComponent.cs

```diff
@@ -0,0 +1,367 @@
+                        EjectSolid(msg.EntityID);
+                        ClickSound();
+                        _dirty = true;
+                    }
+                    break;
+
+                case ReagentGrinderEjectBeakerMessage msg:
+                    ClickSound();
+                    EjectBeaker(message.Session.AttachedEntity!);
+                    //EjectBeaker will dirty the UI for us, we don't have to do it explicitly here.
+                    break;
+            }
+        }
+        private void ClickSound()
+        {
+            _audioSystem.PlayFromEntity("/Audio/Machines/machine_switch.ogg", Owner, AudioParams.Default.WithVolume(-2f));
+        }
+        private void SetAppearance(ReagentGrinderVisualState state)
+        {
+            if (Owner.TryGetComponent(out AppearanceComponent? appearance))
+            {
+                appearance.SetData(PowerDeviceVisuals.VisualState, state);
+            }
+        }
+
+        public void OnUpdate()
+        {
+            if(_dirty)
+            {
+                UserInterface?.SetState(new ReagentGrinderInterfaceState
+                (
+                    _busy,
+                    HasBeaker,
+                    _chamber.ContainedEntities.Select(item => item.Uid).ToArray(),
+                    //Remember the beaker can be null!
+                    _heldBeaker?.Solution.Contents.ToArray()
+                ));
+                _dirty = false;
+            }
+        }
+
+        private void EjectSolid(EntityUid entityID)
+        {
+            if (_busy)
+                return;
+
+            if (_entityManager.EntityExists(entityID))
+            {
+                var entity = _entityManager.GetEntity(entityID);
+                _chamber.Remove(entity);
+
+                //Give the ejected entity a tiny bit of offset so each one is apparent in case of a big stack,
+                //but (hopefully) not enough to clip it through a solid (wall).
+                const float ejectOffset = 0.4f;
+                entity.Transform.LocalPosition += (_random.NextFloat() * ejectOffset, _random.NextFloat() * ejectOffset);
+            }
+            _dirty = true;
+        }
+
+        /// <summary>
+        /// If this component contains an entity with a <see cref="SolutionContainerComponent"/>, eject it.
+        /// Tries to eject into user's hands first, then ejects onto dispenser if both hands are full.
+        /// </summary>
+        private void EjectBeaker(IEntity user)
+        {
+            if (!HasBeaker || _busy)
+                return;
+
+            //Eject the beaker into the hands of the user.
+            _beakerContainer.Remove(_beakerContainer.ContainedEntity);
+
+            //UpdateUserInterface();
+
+            if (!user.TryGetComponent<HandsComponent>(out var hands) || !_heldBeaker!.Owner.TryGetComponent<ItemComponent>(out var item))
+                return;
+            if (hands.CanPutInHand(item))
+                hands.PutInHand(item);
```

Comment:
There is a PutInHandOrDrop method now for this kind of functionality.

Reply:
only took u guys 8 months (probably)